src/collector/rsync.rs

```diff
@@ -445,6 +332,75 @@ impl<'a> Run<'a> {
         }
     }
 
+    /// Cleans the collector only keeping the modules included in `retain`.
+    //
+    //  This currently is super agressive, deleting everyting that it doesn’t
+    //  like.
+    pub fn cleanup(&self, retain: &mut ModuleSet) -> Result<(), Failed> {
+        if self.collector.command.is_none() {
+            return Ok(())
+        }
+
+        // Add all modules we’ve used during this run to retain.
+        for module in self.updated.read().unwrap().iter() {
+            retain.add_from_uri(&module.to_uri());
+        }
+
+        for entry in fatal::read_dir(
+            &self.collector.working_dir.base
+        )? {
+            let entry = entry?;
+            let keep = match entry.file_name().to_str() {
+                Some(name) => {
+                    match retain.authorities.get(name) {
+                        Some(modules) => self.cleanup_host(&entry, modules)?,
+                        None => false,
+                    }
+                }
+                None => false
+            };
+
+            if !keep {
+                fatal::remove_dir_all(entry.path())?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Removes all modules from the directory that are not in `retain`.
+    ///
+    /// Returns whether the host should be kept or can be deleted, too.
+    fn cleanup_host(
+        &self, entry: &fatal::DirEntry, retain: &HashSet<String>
+    ) -> Result<bool, Failed> {
+        if entry.is_file() {
+            fatal::remove_file(entry.path())?;
+            return Ok(false)
+        }
+        else if !entry.is_dir() {
```

Comment:
There may be other things (named pipes, Unix socket, …) which I don’t want to try delete directly because that may or may not fail. They will be deleted with the `remove_dir_all` one level up if necessary (which also may fail, but yeah.)